File: FriendlyTerminal/OutputRendering/Detectors.swift
import Foundation

enum CatImageDetector: OutputDetector {
    static let priority = 5

    private static let imageExtensions: Set<String> = [
        "png", "jpg", "jpeg", "gif", "webp", "heic", "heif", "tiff", "tif", "bmp"
    ]

    static func detect(output: String, command: String, cwd: String) -> RenderKind? {
        let parts = command.trimmingCharacters(in: .whitespaces)
            .components(separatedBy: .whitespaces)
            .filter { !$0.isEmpty }
        guard parts.first == "cat", parts.count >= 2 else { return nil }

        let filePath = parts.last!
        let ext = (filePath as NSString).pathExtension.lowercased()
        guard imageExtensions.contains(ext) else { return nil }

        let url: URL
        if filePath.hasPrefix("/") {
            url = URL(fileURLWithPath: filePath)
        } else if filePath.hasPrefix("~") {
            url = URL(fileURLWithPath: NSString(string: filePath).expandingTildeInPath)
        } else {
            url = URL(fileURLWithPath: cwd, isDirectory: true).appendingPathComponent(filePath)
        }

        guard FileManager.default.fileExists(atPath: url.path) else { return nil }
        return .imageFile(url)
    }
}

enum ImagePathDetector: OutputDetector {
    static let priority = 10

    private static let imageExtensions = Set([
        "png", "jpg", "jpeg", "gif", "webp", "heic", "heif", "tiff", "tif", "bmp", "svg"
    ])

    static func detect(output: String, command: String, cwd: String) -> RenderKind? {
        let lines = output.components(separatedBy: .newlines)
            .map { $0.trimmingCharacters(in: .whitespaces) }
            .filter { !$0.isEmpty }

        guard let lastLine = lines.last else { return nil }

        guard lastLine.hasPrefix("/") else { return nil }
        let ext = (lastLine as NSString).pathExtension.lowercased()
        guard imageExtensions.contains(ext) else { return nil }

        guard FileManager.default.fileExists(atPath: lastLine) else { return nil }

        return .imageFile(URL(fileURLWithPath: lastLine))
    }
}

enum JSONDetector: OutputDetector {
    static let priority = 20

    static func detect(output: String, command: String, cwd: String) -> RenderKind? {
        let trimmed = output.trimmingCharacters(in: .whitespacesAndNewlines)

        guard trimmed.hasPrefix("{") || trimmed.hasPrefix("[") else { return nil }

        guard trimmed.components(separatedBy: .newlines).count >= 2 else { return nil }

        guard let data = trimmed.data(using: .utf8),
              let _ = try? JSONSerialization.jsonObject(with: data) else { return nil }

        return .jsonTree
    }
}

enum CSVDetector: OutputDetector {
    static let priority = 25

    static func detect(output: String, command: String, cwd: String) -> RenderKind? {
        let lines = output.components(separatedBy: .newlines).filter { !$0.isEmpty }
        guard lines.count >= 2 else { return nil }

        let firstLine = lines[0]

        let commas = firstLine.filter { $0 == "," }.count
        let tabs   = firstLine.filter { $0 == "\t" }.count
        let sep: Character
        if commas >= 2 { sep = "," }
        else if tabs >= 1 { sep = "\t" }
        else { return nil }

        let expectedCols = firstLine.split(separator: sep, omittingEmptySubsequences: false).count
        guard expectedCols >= 2 else { return nil }

        let sample = min(lines.count, 20)
        let consistent = lines.prefix(sample).filter {
            $0.split(separator: sep, omittingEmptySubsequences: false).count == expectedCols
        }.count
        guard Double(consistent) / Double(sample) >= 0.8 else { return nil }

        let rows = lines.map {
            $0.split(separator: sep, omittingEmptySubsequences: false).map {
                $0.trimmingCharacters(in: .whitespaces)
                  .trimmingCharacters(in: CharacterSet(charactersIn: "\""))
            }
        }
        return .csvTable(rows)
    }
}

enum TableDetector: OutputDetector {
    static let priority = 30

    static let minCols = 3
    static let minRows = 3
    static let maxRows = 500

    static func detect(output: String, command: String, cwd: String) -> RenderKind? {
        let lines = output.components(separatedBy: .newlines)
            .filter { !$0.trimmingCharacters(in: .whitespaces).isEmpty }

        guard lines.count >= minRows else { return nil }

        let rows = lines.prefix(maxRows).map { line -> [String] in
            splitIntoColumns(line)
        }

        let colCounts = rows.map(\.count)
        let medianCols = colCounts.sorted()[colCounts.count / 2]
        guard medianCols >= minCols else { return nil }

        let consistent = colCounts.filter { abs($0 - medianCols) <= 2 }.count
        guard Double(consistent) / Double(colCounts.count) > 0.7 else { return nil }

        return .table(rows.map { Array($0) })
    }

    private static func splitIntoColumns(_ line: String) -> [String] {
        let pattern = "  +"
        if let regex = try? NSRegularExpression(pattern: pattern) {
            let range = NSRange(line.startIndex..., in: line)
            let result = regex.stringByReplacingMatches(in: line, range: range, withTemplate: "\t")
            return result.components(separatedBy: "\t")
                .map { $0.trimmingCharacters(in: .whitespaces) }
                .filter { !$0.isEmpty }
        }
        return line.components(separatedBy: .whitespaces).filter { !$0.isEmpty }
    }
}

/// Wraps a name in single quotes (escaping any embedded quotes) so it's safe to
/// paste into the command bar even when it contains spaces or special characters.
private func shellQuote(_ name: String) -> String {
    "'" + name.replacingOccurrences(of: "'", with: "'\\''") + "'"
}

/// Picks an SF Symbol for a file based on its extension (folders are handled
/// separately). Mirrors the sidebar's iconography.
private func fileIcon(for name: String) -> String {
    switch (name as NSString).pathExtension.lowercased() {
    case "swift": return "swift"
    case "js", "ts", "jsx", "tsx": return "curlybraces"
    case "json", "yaml", "yml", "toml": return "doc.text"
    case "md", "txt": return "doc.plaintext"
    case "png", "jpg", "jpeg", "gif", "webp", "heic": return "photo"
    case "pdf": return "doc.richtext"
    case "zip", "gz", "tar", "xz": return "archivebox"
    case "sh", "zsh", "bash", "py": return "terminal"
    default: return "doc"
    }
}

/// Turns a plain `ls` of the current folder into clickable entries: folders
/// suggest `cd folder`, files suggest `open file`. Only handles listings of the
/// current directory (no path argument, not recursive); anything else falls
/// through to normal text rendering. Uses the live filesystem for accurate
/// folder-vs-file detection rather than parsing ls's columns.
enum LsListingDetector: OutputDetector {
    static func detect(output: String, command: String, cwd: String) -> RenderKind? {
        let parts = command.trimmingCharacters(in: .whitespaces)
            .components(separatedBy: .whitespaces)
            .filter { !$0.isEmpty }
        guard parts.first == "ls" else { return nil }

        // Every extra argument must be an option flag — a path operand means the
        // listing isn't of `cwd`, so we can't map names to clickable actions.
        let operands = parts.dropFirst()
        for arg in operands {
            guard arg.hasPrefix("-") else { return nil }
            if arg.contains("R") { return nil } // recursive listing — skip
        }
        let showHidden = operands.contains { $0.contains("a") || $0.contains("A") }

        let dir = URL(fileURLWithPath: cwd, isDirectory: true)
        guard let entries = try? FileManager.default.contentsOfDirectory(
            at: dir,
            includingPropertiesForKeys: [.isDirectoryKey, .isHiddenKey],
            options: []
        ) else { return nil }

        var items: [CommandListItem] = []
        for url in entries {
            let values = try? url.resourceValues(forKeys: [.isDirectoryKey, .isHiddenKey])
            let isDir = values?.isDirectory ?? false
            let isHidden = values?.isHidden ?? url.lastPathComponent.hasPrefix(".")
            if isHidden && !showHidden { continue }

            let name = url.lastPathComponent
            items.append(CommandListItem(
                label: name,
                detail: isDir ? "Open this folder (cd)" : "Open this file",
                systemImage: isDir ? "folder.fill" : fileIcon(for: name),
                followUp: isDir ? "cd \(shellQuote(name))" : "open \(shellQuote(name))"
            ))
        }
        guard !items.isEmpty else { return nil }

        items.sort { lhs, rhs in
            let lhsDir = lhs.systemImage == "folder.fill"
            let rhsDir = rhs.systemImage == "folder.fill"
            if lhsDir != rhsDir { return lhsDir }
            return lhs.label.localizedCaseInsensitiveCompare(rhs.label) == .orderedAscending
        }
        return .commandList(
            hint: "Click a folder to fill in “cd” (go into it) or a file to fill in “open” — then press Return.",
            items: items
        )
    }
}

/// Turns `git branch` into clickable branches: tapping one suggests
/// `git checkout branch`. Skips management forms (`-d`, `-m`, naming a branch).
enum GitBranchDetector: OutputDetector {
    static func detect(output: String, command: String, cwd: String) -> RenderKind? {
        let trimmed = command.trimmingCharacters(in: .whitespaces)
        guard trimmed == "git branch" || trimmed.hasPrefix("git branch ") else { return nil }

        let rest = trimmed.dropFirst("git branch".count).trimmingCharacters(in: .whitespaces)
        if !rest.isEmpty {
            // Allow only listing-style flags (e.g. -a, -vv); bail on a branch
            // name operand or destructive flags.
            for arg in rest.components(separatedBy: .whitespaces) where !arg.isEmpty {
                guard arg.hasPrefix("-") else { return nil }
                if arg.contains("d") || arg.contains("D") || arg.contains("m") { return nil }
            }
        }

        var items: [CommandListItem] = []
        for rawLine in output.components(separatedBy: .newlines) {
            var line = rawLine.trimmingCharacters(in: .whitespaces)
            guard !line.isEmpty else { continue }

            let isCurrent = line.hasPrefix("* ")
            if isCurrent || line.hasPrefix("+ ") {
                line = String(line.dropFirst(2)).trimmingCharacters(in: .whitespaces)
            }
            let name = line.components(separatedBy: .whitespaces).first ?? line
            // Skip detached-HEAD lines and tracking arrows.
            if name.isEmpty || name.hasPrefix("(") || name.contains("->") { continue }

            items.append(CommandListItem(
                label: isCurrent ? "\(name)  ●" : name,
                detail: isCurrent ? "Current branch" : "Switch to this branch",
                systemImage: "arrow.triangle.branch",
                followUp: "git checkout \(shellQuote(name))"
            ))
        }
        guard !items.isEmpty else { return nil }
        return .commandList(
            hint: "Click a branch to fill in “git checkout” — then press Return to switch to it.",
            items: items
        )
    }
}

/// Turns `git status` into clickable changed files: tapping one suggests
/// `git add file` to stage it. Handles the default (verbose) output; the short
/// / porcelain forms fall through to text.
enum GitStatusDetector: OutputDetector {
    private static let statusKeywords = [
        "modified:", "new file:", "deleted:", "renamed:", "copied:",
        "typechange:", "both modified:", "both added:", "added by us:",
        "deleted by us:", "added by them:", "deleted by them:", "unmerged:"
    ]

    static func detect(output: String, command: String, cwd: String) -> RenderKind? {
        let trimmed = command.trimmingCharacters(in: .whitespaces)
        guard trimmed == "git status" || trimmed.hasPrefix("git status ") else { return nil }
        // Short/porcelain output has a different shape — leave it as text.
        if trimmed.contains("-s") || trimmed.contains("--short") || trimmed.contains("--porcelain") {
            return nil
        }

        var items: [CommandListItem] = []
        var seen = Set<String>()
        for rawLine in output.components(separatedBy: .newlines) {
            let line = rawLine.trimmingCharacters(in: .whitespaces)
            guard !line.isEmpty else { continue }

            var path: String?
            if let keyword = statusKeywords.first(where: { line.hasPrefix($0) }) {
                path = String(line.dropFirst(keyword.count)).trimmingCharacters(in: .whitespaces)
            } else if rawLine.hasPrefix("\t") && !line.hasPrefix("(") {
                path = line // untracked file (indented, no status keyword)
            }

            guard var file = path, !file.isEmpty else { continue }
            if let arrow = file.range(of: "-> ") { // "renamed: old -> new"
                file = String(file[arrow.upperBound...]).trimmingCharacters(in: .whitespaces)
            }
            guard seen.insert(file).inserted else { continue }

            items.append(CommandListItem(
                label: file,
                detail: "Stage this file (git add)",
                systemImage: "plus.circle",
                followUp: "git add \(shellQuote(file))"
            ))
        }
        guard !items.isEmpty else { return nil }
        return .commandList(
            hint: "Click a file to fill in “git add” for it — then press Return to stage it.",
            items: items
        )
    }
}

/// Turns `git log --oneline` into clickable commits: tapping one suggests
/// `git show <hash>`.
enum GitLogOnelineDetector: OutputDetector {
    static func detect(output: String, command: String, cwd: String) -> RenderKind? {
        let trimmed = command.trimmingCharacters(in: .whitespaces)
        guard trimmed.hasPrefix("git log"), trimmed.contains("--oneline") else { return nil }

        var items: [CommandListItem] = []
        for rawLine in output.components(separatedBy: .newlines) {
            let line = rawLine.trimmingCharacters(in: .whitespaces)
            guard !line.isEmpty else { continue }
            let hash = String(line.split(separator: " ", maxSplits: 1).first ?? "")
            guard hash.count >= 4, hash.allSatisfy(\.isHexDigit) else { continue }
            items.append(CommandListItem(
                label: line,
                detail: "Show this commit (git show)",
                systemImage: "clock",
                followUp: "git show \(hash)"
            ))
        }
        guard !items.isEmpty else { return nil }
        return .commandList(
            hint: "Click a commit to fill in “git show” — then press Return to view it.",
            items: items
        )
    }
}

/// Turns `git tag` into clickable tags: tapping one suggests `git checkout tag`.
enum GitTagDetector: OutputDetector {
    static func detect(output: String, command: String, cwd: String) -> RenderKind? {
        let trimmed = command.trimmingCharacters(in: .whitespaces)
        guard trimmed == "git tag"
            || trimmed.hasPrefix("git tag -l")
            || trimmed.hasPrefix("git tag --list") else { return nil }

        var items: [CommandListItem] = []
        for rawLine in output.components(separatedBy: .newlines) {
            let line = rawLine.trimmingCharacters(in: .whitespaces)
            guard !line.isEmpty, !line.contains(" ") else { continue }
            items.append(CommandListItem(
                label: line,
                detail: "Check out this tag",
                systemImage: "tag",
                followUp: "git checkout \(shellQuote(line))"
            ))
        }
        guard !items.isEmpty else { return nil }
        return .commandList(
            hint: "Click a tag to fill in “git checkout” — then press Return to check it out.",
            items: items
        )
    }
}

/// Turns `brew list` into clickable packages: tapping one suggests `brew info`.
enum BrewListDetector: OutputDetector {
    static func detect(output: String, command: String, cwd: String) -> RenderKind? {
        let parts = command.trimmingCharacters(in: .whitespaces)
            .components(separatedBy: .whitespaces)
            .filter { !$0.isEmpty }
        guard parts.first == "brew",
              let sub = parts.dropFirst().first, (sub == "list" || sub == "ls") else { return nil }
        // "--versions" prints "name version" pairs — a different shape.
        if parts.contains("--versions") { return nil }

        var items: [CommandListItem] = []
        var seen = Set<String>()
        for rawLine in output.components(separatedBy: .newlines) {
            let line = rawLine.trimmingCharacters(in: .whitespaces)
            if line.isEmpty || line.hasPrefix("==>") { continue }
            for token in line.components(separatedBy: .whitespaces) where !token.isEmpty {
                guard seen.insert(token).inserted else { continue }
                items.append(CommandListItem(
                    label: token,
                    detail: "See package details (brew info)",
                    systemImage: "shippingbox",
                    followUp: "brew info \(shellQuote(token))"
                ))
            }
        }
        guard !items.isEmpty else { return nil }
        return .commandList(
            hint: "Click a package to fill in “brew info” — then press Return to see its details.",
            items: items
        )
    }
}

/// Turns `history` into clickable past commands: tapping one drops that command
/// back into the command bar, ready to run or tweak.
enum HistoryDetector: OutputDetector {
    static func detect(output: String, command: String, cwd: String) -> RenderKind? {
        let trimmed = command.trimmingCharacters(in: .whitespaces)
        guard trimmed == "history" || trimmed.hasPrefix("history ") else { return nil }

        var items: [CommandListItem] = []
        for rawLine in output.components(separatedBy: .newlines) {
            let line = rawLine.trimmingCharacters(in: .whitespaces)
            guard !line.isEmpty else { continue }
            // Lines look like "  123  the command".
            let pieces = line.split(separator: " ", maxSplits: 1, omittingEmptySubsequences: true)
            guard pieces.count == 2, Int(pieces[0]) != nil else { continue }
            let past = pieces[1].trimmingCharacters(in: .whitespaces)
            guard !past.isEmpty else { continue }
            items.append(CommandListItem(
                label: past,
                detail: "Use this command again",
                systemImage: "clock.arrow.circlepath",
                followUp: past
            ))
        }
        guard !items.isEmpty else { return nil }
        if items.count > 200 { items = Array(items.suffix(200)) }
        return .commandList(
            hint: "Click a past command to load it into the command bar — then press Return to run it again.",
            items: items
        )
    }
}

enum FileTreeDetector: OutputDetector {
    static let priority = 40

    static func detect(output: String, command: String, cwd: String) -> RenderKind? {
        if output.contains("├──") || output.contains("└──") || output.contains("│") {
            return .fileTree
        }
        let lines = output.components(separatedBy: .newlines).filter { !$0.isEmpty }
        guard lines.count >= 3 else { return nil }
        let pathLines = lines.filter { $0.hasPrefix("./") || ($0.hasPrefix("/") && $0.contains("/")) }
        if Double(pathLines.count) / Double(lines.count) > 0.8 {
            return .fileTree
        }
        return nil
    }
}

File: windows/src/FriendlyTerminal.Core/Output/Detectors.cs
using System.Text.Json;
using System.Text.RegularExpressions;
using FriendlyTerminal.Core.Platform;

namespace FriendlyTerminal.Core.Output;

internal static class DetectorUtil
{
    public static readonly HashSet<string> ImageExtensions = new(StringComparer.OrdinalIgnoreCase)
    {
        ".png", ".jpg", ".jpeg", ".gif", ".webp", ".bmp", ".ico", ".tiff", ".tif"
    };

    public static string ResolvePath(string p, string cwd, string home)
    {
        if (p.StartsWith('~')) return home + p[1..];
        if (IsRooted(p)) return p;
        // Join with '/' (Windows APIs accept both separators) so fake and real
        // file systems agree on the shape.
        return cwd.TrimEnd('/', '\\') + "/" + p;
    }

    /// <summary>
    /// Windows-style rootedness, independent of the host OS. Path.IsPathRooted
    /// rejects "C:\..." when running on Linux, which broke these detectors on
    /// the Linux CI runner.
    /// </summary>
    public static bool IsRooted(string path) =>
        (path.Length >= 3 && char.IsAsciiLetter(path[0]) && path[1] == ':' && (path[2] == '\\' || path[2] == '/'))
        || path.StartsWith('\\') || path.StartsWith('/');

    public static string IconForFile(string name) =>
        System.IO.Path.GetExtension(name).ToLowerInvariant() switch
        {
            ".cs" or ".js" or ".ts" or ".jsx" or ".tsx" or ".py" or ".ps1" or ".sh" or ".bat" or ".cmd" => "code",
            ".json" or ".yaml" or ".yml" or ".toml" => "doc",
            ".md" or ".txt" => "doc",
            ".png" or ".jpg" or ".jpeg" or ".gif" or ".webp" or ".bmp" or ".ico" => "picture",
            ".pdf" => "pdf",
            ".zip" or ".gz" or ".tar" or ".7z" or ".rar" => "archive",
            _ => "file",
        };
}

/// <summary>
/// Renders `ls` / `dir` / `Get-ChildItem` (no target argument) as clickable chips
/// built from the real directory contents, matching the macOS LsListingDetector.
/// </summary>
public sealed class LsListingDetector : IOutputDetector
{
    private static readonly HashSet<string> Commands = new(StringComparer.OrdinalIgnoreCase)
    {
        "ls", "dir", "gci", "Get-ChildItem"
    };

    private readonly IFileSystem _fs;
    private readonly IShellQuoter _quoter;

    public LsListingDetector(IFileSystem fs, IShellQuoter quoter)
    {
        _fs = fs;
        _quoter = quoter;
    }

    public RenderKind? Detect(string output, string command, string cwd)
    {
        var parts = command.Trim().Split(' ', '\t').Where(p => p.Length > 0).ToArray();
        if (parts.Length == 0 || !Commands.Contains(parts[0])) return null;

        var showHidden = false;
        foreach (var arg in parts.Skip(1))
        {
            if (!arg.StartsWith('-')) return null; // listing another folder - fall through
            if (arg.Contains("Recurse", StringComparison.OrdinalIgnoreCase) || arg.Contains('R')) return null;
            if (arg.Contains("Force", StringComparison.OrdinalIgnoreCase) || arg.Contains('a')) showHidden = true;
        }

        var items = new List<CommandListItem>();
        foreach (var entry in _fs.ListEntries(cwd))
        {
            if (entry.IsHidden && !showHidden) continue;
            items.Add(new CommandListItem(
                Label: entry.Name,
                Detail: entry.IsDirectory ? "Open this folder (cd)" : "Open this file",
                Icon: entry.IsDirectory ? "folder" : DetectorUtil.IconForFile(entry.Name),
                FollowUp: entry.IsDirectory
                    ? $"cd {_quoter.Quote(entry.Name)}"
                    : $"Invoke-Item {_quoter.Quote(entry.Name)}"));
        }
        if (items.Count == 0) return null;

        items.Sort((a, b) =>
        {
            var aDir = a.Icon == "folder";
            var bDir = b.Icon == "folder";
            if (aDir != bDir) return aDir ? -1 : 1;
            return string.Compare(a.Label, b.Label, StringComparison.OrdinalIgnoreCase);
        });

        return new RenderKind.CommandList(
            "Click a folder to fill in “cd” (go into it) or a file to open it — then press Enter.",
            items);
    }
}

public sealed class GitLogOnelineDetector : IOutputDetector
{
    public RenderKind? Detect(string output, string command, string cwd)
    {
        var trimmed = command.Trim();
        if (!trimmed.StartsWith("git log") || !trimmed.Contains("--oneline")) return null;

        var items = new List<CommandListItem>();
        foreach (var rawLine in output.Split('\n'))
        {
            var line = rawLine.Trim();
            if (line.Length == 0) continue;
            var hash = line.Split(' ', 2)[0];
            if (hash.Length < 4 || !hash.All(Uri.IsHexDigit)) continue;
            items.Add(new CommandListItem(
                Label: line,
                Detail: "Show this commit (git show)",
                Icon: "clock",
                FollowUp: $"git show {hash}"));
        }
        if (items.Count == 0) return null;
        return new RenderKind.CommandList(
            "Click a commit to fill in “git show” — then press Enter to view it.",
            items);
    }
}

public sealed class GitTagDetector : IOutputDetector
{
    private readonly IShellQuoter _quoter;

    public GitTagDetector(IShellQuoter quoter) => _quoter = quoter;

    public RenderKind? Detect(string output, string command, string cwd)
    {
        var trimmed = command.Trim();
        if (trimmed != "git tag" && !trimmed.StartsWith("git tag -l") && !trimmed.StartsWith("git tag --list"))
            return null;

        var items = new List<CommandListItem>();
        foreach (var rawLine in output.Split('\n'))
        {
            var line = rawLine.Trim();
            if (line.Length == 0 || line.Contains(' ')) continue;
            items.Add(new CommandListItem(
                Label: line,
                Detail: "Check out this tag",
                Icon: "tag",
                FollowUp: $"git checkout {_quoter.Quote(line)}"));
        }
        if (items.Count == 0) return null;
        return new RenderKind.CommandList(
            "Click a tag to fill in “git checkout” — then press Enter to check it out.",
            items);
    }
}

/// <summary>
/// PowerShell `history` / `Get-History` output: an Id column followed by the
/// command line. Clicking a row loads that command into the command bar.
/// </summary>
public sealed class HistoryDetector : IOutputDetector
{
    private static readonly HashSet<string> Commands = new(StringComparer.OrdinalIgnoreCase)
    {
        "history", "h", "Get-History"
    };

    public RenderKind? Detect(string output, string command, string cwd)
    {
        var first = command.Trim().Split(' ', 2)[0];
        if (!Commands.Contains(first)) return null;

        var items = new List<CommandListItem>();
        foreach (var rawLine in output.Split('\n'))
        {
            var line = rawLine.Trim();
            if (line.Length == 0) continue;
            var pieces = line.Split(' ', 2, StringSplitOptions.RemoveEmptyEntries);
            if (pieces.Length != 2 || !int.TryParse(pieces[0], out _)) continue;
            var past = pieces[1].Trim();
            if (past.Length == 0) continue;
            items.Add(new CommandListItem(
                Label: past,
                Detail: "Use this command again",
                Icon: "history",
                FollowUp: past));
        }
        if (items.Count == 0) return null;
        if (items.Count > 200) items = items.Skip(items.Count - 200).ToList();
        return new RenderKind.CommandList(
            "Click a past command to load it into the command bar — then press Enter to run it again.",
            items);
    }
}

/// <summary>`cat` / `Get-Content` on an image file: show the picture instead of bytes.</summary>
public sealed class CatImageDetector : IOutputDetector
{
    private static readonly HashSet<string> Commands = new(StringComparer.OrdinalIgnoreCase)
    {
        "cat", "gc", "type", "Get-Content"
    };

    private readonly IFileSystem _fs;

    public CatImageDetector(IFileSystem fs) => _fs = fs;

    public RenderKind? Detect(string output, string command, string cwd)
    {
        var parts = command.Trim().Split(' ', '\t').Where(p => p.Length > 0).ToArray();
        if (parts.Length < 2 || !Commands.Contains(parts[0])) return null;

        var filePath = parts[^1];
        if (!DetectorUtil.ImageExtensions.Contains(System.IO.Path.GetExtension(filePath))) return null;

        var resolved = DetectorUtil.ResolvePath(filePath, cwd, _fs.HomeDirectory);
        return _fs.Exists(resolved) ? new RenderKind.ImageFile(resolved) : null;
    }
}

/// <summary>Output whose last line is an absolute path to an existing image: show it.</summary>
public sealed class ImagePathDetector : IOutputDetector
{
    private readonly IFileSystem _fs;

    public ImagePathDetector(IFileSystem fs) => _fs = fs;

    public RenderKind? Detect(string output, string command, string cwd)
    {
        var lastLine = output.Split('\n')
            .Select(l => l.Trim())
            .LastOrDefault(l => l.Length > 0);
        if (lastLine is null || !DetectorUtil.IsRooted(lastLine)) return null;
        if (!DetectorUtil.ImageExtensions.Contains(System.IO.Path.GetExtension(lastLine))) return null;
        return _fs.Exists(lastLine) ? new RenderKind.ImageFile(lastLine) : null;
    }
}

public sealed class JsonDetector : IOutputDetector
{
    public RenderKind? Detect(string output, string command, string cwd)
    {
        var trimmed = output.Trim();
        if (!trimmed.StartsWith('{') && !trimmed.StartsWith('[')) return null;
        if (trimmed.Split('\n').Length < 2) return null;
        try
        {
            using var _ = JsonDocument.Parse(trimmed);
            return new RenderKind.JsonTree();
        }
        catch (JsonException)
        {
            return null;
        }
    }
}

public sealed class CsvDetector : IOutputDetector
{
    public RenderKind? Detect(string output, string command, string cwd)
    {
        var lines = output.Split('\n').Where(l => l.Length > 0).ToArray();
        if (lines.Length < 2) return null;

        var firstLine = lines[0];
        var commas = firstLine.Count(c => c == ',');
        var tabs = firstLine.Count(c => c == '\t');
        char sep;
        if (commas >= 2) sep = ',';
        else if (tabs >= 1) sep = '\t';
        else return null;

        var expectedCols = firstLine.Split(sep).Length;
        if (expectedCols < 2) return null;

        var sample = Math.Min(lines.Length, 20);
        var consistent = lines.Take(sample).Count(l => l.Split(sep).Length == expectedCols);
        if ((double)consistent / sample < 0.8) return null;

        var rows = lines
            .Select(l => l.Split(sep).Select(c => c.Trim().Trim('"')).ToArray())
            .ToList();
        return new RenderKind.CsvTable(rows);
    }
}

public sealed class TableDetector : IOutputDetector
{
    private const int MinCols = 3;
    private const int MinRows = 3;
    private const int MaxRows = 500;

    private static readonly Regex MultiSpace = new("  +", RegexOptions.Compiled);

    public RenderKind? Detect(string output, string command, string cwd)
    {
        var lines = output.Split('\n').Where(l => l.Trim().Length > 0).ToArray();
        if (lines.Length < MinRows) return null;

        var rows = lines.Take(MaxRows).Select(SplitIntoColumns).ToList();

        var colCounts = rows.Select(r => r.Length).ToArray();
        var medianCols = colCounts.OrderBy(c => c).ElementAt(colCounts.Length / 2);
        if (medianCols < MinCols) return null;

        var consistent = colCounts.Count(c => Math.Abs(c - medianCols) <= 2);
        if ((double)consistent / colCounts.Length <= 0.7) return null;

        return new RenderKind.Table(rows);
    }

    private static string[] SplitIntoColumns(string line) =>
        MultiSpace.Replace(line, "\t")
            .Split('\t')
            .Select(c => c.Trim())
            .Where(c => c.Length > 0)
            .ToArray();
}
